File: src/main/java/com/github/thorben/webnd/foundation/AbstractValueObject.java
package com.github.thorben.webnd.foundation;

import javax.persistence.Access;
import javax.persistence.AccessType;
import javax.persistence.Column;
import javax.persistence.MappedSuperclass;
import java.io.Serializable;

import static org.springframework.util.Assert.notNull;

@MappedSuperclass
@Access(AccessType.FIELD)
public abstract class AbstractValueObject<V extends Comparable<? super V>>
		implements Comparable<AbstractValueObject<V>>, Serializable {

	private static final String VALUE = "value";
	@Column(name = VALUE)
	private V value;

	protected AbstractValueObject() {
		// For JPA
	}

	public AbstractValueObject(V value) {
		notNull(value);
		this.value = value;
	}

	public V getValue() {
		return value;
	}

	@Override
	public int compareTo(AbstractValueObject<V> vAbstractValueObject) {
		return getValue().compareTo(vAbstractValueObject.getValue());
	}

	@Override
	public String toString() {
		return getValue().toString();
	}

	@Override
	public int hashCode() {
		return getValue().hashCode();
	}

	@Override
	public boolean equals(Object that) {
		if (this == that) {
			return true;
		}

		if (!(that instanceof AbstractValueObject) ||
				!(that.getClass().isAssignableFrom(getClass())) ||
				getClass().isAssignableFrom(that.getClass())) {
			return false;
		}

		AbstractValueObject<V> valueObject = (AbstractValueObject<V>) that;
		return valueObject.getValue().equals(getValue());
	}
}

File: src/main/kotlin/com/github/thorben/webnd/foundation/AbstractValueObject.kt
package com.github.thorben.webnd.foundation

import java.io.Serializable
import javax.persistence.Access
import javax.persistence.AccessType
import javax.persistence.Column
import javax.persistence.MappedSuperclass

@MappedSuperclass
@Access(AccessType.FIELD)
abstract class AbstractValueObject<V : Any> : Serializable {
	@Column(name = VALUE)
	lateinit var value: V
		private set

	protected constructor() {
		// For JPA
	}

	constructor(value: V) {
		this.value = value
	}

	override fun toString(): String {
		return value.toString()
	}

	override fun hashCode(): Int {
		return value.hashCode()
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) {
			return true
		}
		if(other == null) {
			return false
		}
		if (other !is AbstractValueObject<*> ||
				!other.javaClass.isAssignableFrom(javaClass) ||
				javaClass.isAssignableFrom(other.javaClass)) {
			return false
		}
		val valueObject = other as AbstractValueObject<V>
		return valueObject.value == value
	}

	companion object {
		private const val VALUE = "value"
	}
}